### DISCUSSION TITLE
`for (int` on STL containers or loop termination value of type `size_t`

### DISCUSSION THREAD

**Discussion by Naktakala** [chi-tech/chi-tech#151] [General]
So we've all run into this `int`, `unsigned int`, `size_t` business and I've been scouring style and standards documents to find an answer. Seems there is no answer for situations as shown below:

```c++
std::vector<int> objects{5,4,3,2,1};

// Do you have reasonable confidence in the size of objects?
// faces, vertex_ids, groups = Hell yes!
// cells = Hell no!

for (int i=0; i<objects.size(); ++i)
{
    //stuff
    double a_sciency_thing = M_PI*objects[i]; //RISK: i >> max-int
    double b_sciency_thing = M_PI*(-i);
    //more stuff
}

int64_t num_objects = static_cast<int64_t>(objects.size());
for (size_t i=0; i<objects.size(); ++i)
{
    int ii = static_cast<int>(i);
    //stuff
    double a_sciency_thing = M_PI*objects[i]; 
    double b_sciency_thing = M_PI*(-i);        //  RISK: unsigned arithmetic on i
                                                                       //        especially in formulas
    //more stuff
}
```

My opinion:
 - If you have reasonable confidence that the loop iteration counter
   will never overflow then int will remove the risk of unsigned arithmetic
 - Conversely, if you don't know, then use size_t but be cognizant of 
   of unsigned arithmetic

@ragusa @dominiccaron @wdhawkins @jthano

**Comment by Naktakala** [comment]
I thought we resolved that the first case was ok. I get no compiler warning or clang-tidy warnings for these but it seems some do (@wdhawkins).

We need to figure this out because I definitely don't want to rehash this every PR.

**Reply by wdhawkins** [reply]
For whatever reason, clang doesn't add -Wsign-conversion or -Wsign-compare to -Wall.  You have to explicitly enable it or use -Weverything (that's an oversight in my opinion).  If you enable it, you'll get something like:

warning: implicit conversion changes signedness: 'int' to 'std::vector::size_type' (aka 'unsigned long') [-Wsign-conversion]
warning: comparison of integers of different signs: 'int' and 'std::vector::size_type' (aka 'unsigned long') [-Wsign-compare]

I favor avoiding implicit conversions wherever possible.  In the cases where our loop termination condition is a size_t, the loop counter should also be a size_t.

**Reply by dominiccaron** [reply]
I confirm the observations of Daryl regarding the behaviour of clang. On the use of ```-Weverything```, I thought that the flag was intended mainly for clang developers, as it is quite pedantic. That is not to say that it would not be nice to be able to compile cleanly our code with that flag enabled. Should we begin to use it?

Regarding gcc, sign comparison (among other warnings) is explicitly disabled in the current ```CMakeLists.txt``` through the flag ```-Wno-sign-compare```. I have been doing some work to gradually remove these explicitly disabled warnings (and make the corresponding modifications to eliminate error messages). However, I have not even attempted the removal of the ```-Wno-sign-compare``` as I observed that it creates a great number of errors (precisely those mentioned by Daryl). Should we remove it anyway?

**Reply by Naktakala** [reply]
So I thought we settled on this with the following from PR #141 :

From @dominiccaron:
_Merely an observation: I have noticed various concerns arising from inconsistency of types between loop counters and loop limits, but why are there no concerns regarding inconsistency of types arising from the use of the loop counter within the body of the loop (e.g. the signature of the function to which it might be passed)? I mention this as we have many instances where ensuring consistency between any two of the three results in an inconsistency with the third. To me, the ideal is to be entirely consistent, but at the same time, I think that implicit type conversion is handling some of the burden._

From @wdhawkins:
_I suspect Dominic is correct about implicit type conversion. My $0.02 on our signed/unsigned/int/size_t issues. Unless the quantity represented is unlikely to fit into 32 bits, use an int. This is simple, consistent, and natural from a coding perspective. If the quantity will exceed the value that can be represented by 32 bits, using an int64_t or uint64_t (likely only necessary for global ids). IMHO, size_t was an unfortunately choice by the maintainers, but I see why they did it... they really didn't have much choice. Use size_t where needed (looping over container sizes say) and cast to int as necessary._

**Reply by Naktakala** [reply]
Having `for (size_t` everywhere seems sooo gross to me.

**Reply by wdhawkins** [reply]
As I commented in PR #141, I do think size_t was an unfortunate choice by the standard committee, but I it should be used where needed.  It's not necessary to have for(size_t...) everywhere, but I do think it's necessary for looping over containers in that fashion.  Not using size_t is technically a bug that we are depending on the compiler to silently (or not... at least with the proper flags) correct.

**Comment by Naktakala** [comment]
Ok. `-Wsign-conversion` and `-Wsign-compare` will have to be added. And we will have to go through a huge conversion which I will address with an issue and PR. Issue #166 has been created.

I am saddened by this and do not agree that this will lead to fewer risks, but understand the compliance need.
